README.md
# MicroPython DS1307

[![Downloads](https://pepy.tech/badge/micropython-ds1307)](https://pepy.tech/project/micropython-ds1307)
![Release](https://img.shields.io/github/v/release/brainelectronics/micropython-ds1307?include_prereleases&color=success)
![MicroPython](https://img.shields.io/badge/micropython-Ok-green.svg)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![codecov](https://codecov.io/github/brainelectronics/micropython-ds1307/branch/main/graph/badge.svg)](https://app.codecov.io/github/brainelectronics/micropython-ds1307)
[![CI](https://github.com/brainelectronics/micropython-ds1307/actions/workflows/release.yml/badge.svg)](https://github.com/brainelectronics/micropython-ds1307/actions/workflows/release.yml)

MicroPython driver for DS1307 RTC

---------------

## General

MicroPython driver for DS1307 RTC

📚 The latest documentation is available at
[MicroPython DS1307 ReadTheDocs][ref-rtd-micropython-ds1307] 📚

<!-- MarkdownTOC -->

- [Installation](#installation)
	- [Install required tools](#install-required-tools)
- [Setup](#setup)
	- [Install package with upip](#install-package-with-upip)
		- [General](#general)
		- [Specific version](#specific-version)
		- [Test version](#test-version)
	- [Manually](#manually)
		- [Upload files to board](#upload-files-to-board)
- [Usage](#usage)
- [Contributing](#contributing)
	- [Unittests](#unittests)
- [Credits](#credits)

<!-- /MarkdownTOC -->

## Installation

### Install required tools

Python3 must be installed on your system. Check the current Python version
with the following command

```bash
python --version
python3 --version
```

Depending on which command `Python 3.x.y` (with x.y as some numbers) is
returned, use that command to proceed.

```bash
python3 -m venv .venv
source .venv/bin/activate

pip install -r requirements.txt
```

## Setup

### Install package with upip

Connect the MicroPython device to a network (if possible)

```python
import network
station = network.WLAN(network.STA_IF)
station.active(True)
station.connect('SSID', 'PASSWORD')
station.isconnected()
```

#### General

Install the latest package version of this lib on the MicroPython device

```python
import mip
mip.install("github:brainelectronics/micropython-ds1307")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
upip.install('micropython-ds1307')
```

#### Specific version

Install a specific, fixed package version of this lib on the MicroPython device

```python
import mip
# install a verions of a specific branch
mip.install("github:brainelectronics/micropython-ds1307", version="feature/initial-implementation")
# install a tag version
mip.install("github:brainelectronics/micropython-ds1307", version="0.1.0")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
upip.install('micropython-ds1307')
```

#### Test version

Install a specific release candidate version uploaded to
[Test Python Package Index](https://test.pypi.org/) on every PR on the
MicroPython device. If no specific version is set, the latest stable version
will be used.

```python
import mip
mip.install("github:brainelectronics/micropython-ds1307", version="0.1.0-rc1.dev1")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
# overwrite index_urls to only take artifacts from test.pypi.org
upip.index_urls = ['https://test.pypi.org/pypi']
upip.install('micropython-ds1307')
```

See also [brainelectronics Test PyPi Server in Docker][ref-brainelectronics-test-pypiserver]
for a test PyPi server running on Docker.

### Manually

#### Upload files to board

Copy the module to the MicroPython board and import them as shown below
using [Remote MicroPython shell][ref-remote-upy-shell]

Open the remote shell with the following command. Additionally use `-b 115200`
in case no CP210x is used but a CH34x.

```bash
rshell --port /dev/tty.SLAB_USBtoUART --editor nano
```

Perform the following command inside the `rshell` to copy all files and
folders to the device

```bash
mkdir /pyboard/lib
mkdir /pyboard/lib/ds1307

cp ds1307/* /pyboard/lib/ds1307

cp examples/main.py /pyboard
cp examples/boot.py /pyboard
```

## Usage

```python
from ds1307 import DS1307
from machine import I2C, Pin
from time import gmtime, time

# DS1307 on 0x68
I2C_ADDR = 0x68     # DEC 104, HEX 0x68

# define custom I2C interface, default is 'I2C(0)'
# check the docs of your device for further details and pin infos
# this are the pins for the Raspberry Pi Pico adapter board
i2c = I2C(0, scl=Pin(13), sda=Pin(12), freq=800000)
ds1307 = DS1307(addr=I2C_ADDR, i2c=i2c)

# get the current RTC time
print("Current RTC time: {}".format(ds1307.datetime))

# set the RTC time to the current system time
now = gmtime(time())
ds1307.datetime = now

print("Current RTC time: {}".format(ds1307.datetime))

# Print the date and time in ISO8601 format: 2023-04-18T21:14:22
print("Today is {:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}".format(
    ds1307.year, ds1307.month, ds1307.day,
    ds1307.hour, ds1307.minute, ds1307.second))
```

## Contributing

### Unittests

Run the unittests locally with the following command after installing this
package in a virtual environment

```bash
# run all tests
nose2 --config tests/unittest.cfg

# run only one specific tests
nose2 tests.test_ds1307.TestDS1307.test_addr
```

Generate the coverage files with

```bash
python create_report_dirs.py
coverage html
```

The coverage report is placed at `reports/coverage/html/index.html`

## Credits

Based on
[Mike Causer's MicroPython TinyRTC I2C module][ref-micropython-tinyrtc-i2c]
and the [PyPa sample project][ref-pypa-sample]

<!-- Links -->
[ref-rtd-micropython-ds1307]: https://micropython-ds1307.readthedocs.io/en/latest/
[ref-remote-upy-shell]: https://github.com/dhylands/rshell
[ref-brainelectronics-test-pypiserver]: https://github.com/brainelectronics/test-pypiserver
[ref-micropython-tinyrtc-i2c]: https://github.com/mcauser/micropython-tinyrtc-i2c
[ref-pypa-sample]: https://github.com/pypa/sampleproject
[ref-test-pypi]: https://test.pypi.org/
[ref-pypi]: https://pypi.org/


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--
## [x.y.z] - yyyy-mm-dd
### Added
### Changed
### Removed
### Fixed
-->
<!--
RegEx for release version from file
r"^\#\# \[\d{1,}[.]\d{1,}[.]\d{1,}\] \- \d{4}\-\d{2}-\d{2}$"
-->

## Released
## [0.1.1] - 2023-06-12
### Fixed
- Usage documentation with more comments and WiFi instructions in root README
- Ignore version of `package.json` file in test workflow

## [0.1.0] - 2023-05-01
### Added
- `ds1307.py` in `ds1307` module
- Examples and documentation

### Changed
- Several updates on setup and config files different than the template repo

### Removed
- Not used files provided with [template repo](https://github.com/brainelectronics/micropython-i2c-lcd)

<!-- Links -->
[Unreleased]: https://github.com/brainelectronics/micropython-ds1307/compare/0.1.1...main

[0.1.1]: https://github.com/brainelectronics/micropython-ds1307/tree/0.1.1
[0.1.0]: https://github.com/brainelectronics/micropython-ds1307/tree/0.1.0


tests/test_ds1307.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""Unittest for MicroPython DS1307"""

import logging
import sys
import unittest
from unittest.mock import MagicMock, Mock, patch

from nose2.tools import params


class Pin(object):
    """Fake MicroPython Pin class"""
    def __init__(self, pin: int, mode: int = -1):
        self._pin = pin
        self._mode = mode
        self._value = 0


class I2C(object):
    """Fake MicroPython I2C class"""
    def __init__(self, id: int, *, scl: Pin, sda: Pin, freq: int = 400000):
        self._id = id
        self._scl = scl
        self._sda = sda
        self._freq = freq

    def writeto(addr: int, buf: bytearray, stop: bool = True) -> int:
        return 1

    def readfrom_mem(addr: int,
                     memaddr: int,
                     nbytes: int,
                     *,
                     addrsize: int = 8) -> int:
        return 1

    def writeto_mem(addr: int,
                    memaddr: int,
                    buf: bytearray,
                    *,
                    addrsize: int = 8):
        pass


# custom imports
sys.modules['machine.I2C'] = I2C
to_be_mocked = [
    'machine',
    'time.sleep_ms', 'time.sleep_us',
]
for module in to_be_mocked:
    sys.modules[module] = Mock()

from ds1307 import DS1307  # noqa: E402


class TestDS1307(unittest.TestCase):
    """This class describes a TestDS1307 unittest."""

    def setUp(self) -> None:
        """Run before every test method"""
        # define a format
        custom_format = "[%(asctime)s][%(levelname)-8s][%(filename)-20s @" \
                        " %(funcName)-15s:%(lineno)4s] %(message)s"

        # set basic config and level for all loggers
        logging.basicConfig(level=logging.INFO,
                            format=custom_format,
                            stream=sys.stdout)

        # create a logger for this TestSuite
        self.test_logger = logging.getLogger(__name__)

        # set the test logger level
        self.test_logger.setLevel(logging.DEBUG)

        # enable/disable the log output of the device logger for the tests
        # if enabled log data inside this test will be printed
        self.test_logger.disabled = False

        self.i2c = I2C(1, scl=Pin(13), sda=Pin(12), freq=800_000)
        self._calls_counter: int = 0
        self._tracked_call_data: list = []
        self._mocked_read_data: bytes = b''
        self.ds1307 = DS1307()

    def _tracked_call(self, *args, **kwargs) -> None:
        """Track function calls and the used arguments"""
        self._tracked_call_data.append({'args': args, 'kwargs': kwargs})

    def test_init(self) -> None:
        """Test DS1307 init function"""
        ds1307 = DS1307()
        self.assertEqual(ds1307.addr, 0x68)
        # self.assertEqual(ds1307._i2c._id, 0)

        ds1307 = DS1307(i2c=self.i2c)
        self.assertEqual(ds1307.addr, 0x68)
        self.assertEqual(ds1307._i2c._id, 1)

    @params(
        (0x68),
        (104),
    )
    def test_addr(self, addr: int) -> None:
        """Test address property"""
        ds1307 = DS1307(addr=addr)

        self.assertEqual(ds1307.addr, addr)

    def test_weekday_start(self) -> None:
        """Test weekday property"""
        self.assertEqual(self.ds1307.weekday_start, 0)

        # test valid weekday start
        self.ds1307.weekday_start = 1
        self.assertEqual(self.ds1307.weekday_start, 1)

        # test invalid weekday start
        for day in [-1, 7]:
            with self.assertRaises(ValueError) as context:
                self.ds1307.weekday_start = day

            self.assertEqual(
                str(context.exception),
                "Weekday can only be in range 0-6"
            )

    @patch.object(I2C, 'readfrom_mem')
    def test_datetime(self, mock_readfrom_mem: MagicMock) -> None:
        """Test datetime property"""
        ds1307 = DS1307(i2c=self.i2c)

        # 2023-05-29T20:45:59
        #                0,  1,  2, 3,  4,  5, 6
        #                s,  m,  h, wd, d,  m, y
        buf = bytearray([89, 69, 32, 3, 41, 5, 35])
        mock_readfrom_mem.return_value = buf

        self.assertEqual(
            ds1307.datetime,
            (2023, 5, 29, 20, 45, 59, 3, 149)
        )

        with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
            ds1307.datetime = (2023, 5, 29, 20, 45, 59, 3, 149)

            self.assertEqual(len(self._tracked_call_data), 1)
            self.assertEqual(
                self._tracked_call_data[0]['args'],
                (ds1307.addr, 0, buf)
            )

    @patch.object(I2C, 'readfrom_mem')
    def test_year(self, mock_readfrom_mem: MagicMock) -> None:
        """Test year property"""
        ds1307 = DS1307(i2c=self.i2c)

        # 2023-05-29T20:45:59
        #                0,  1,  2, 3,  4,  5, 6
        #                s,  m,  h, wd, d,  m, y
        buf = bytearray([89, 69, 32, 3, 41, 5, 35])
        mock_readfrom_mem.return_value = buf

        self.assertEqual(ds1307.year, 2023)

    @patch.object(I2C, 'readfrom_mem')
    def test_month(self, mock_readfrom_mem: MagicMock) -> None:
        """Test month property"""
        ds1307 = DS1307(i2c=self.i2c)

        # 2023-05-29T20:45:59
        #                0,  1,  2, 3,  4,  5, 6
        #                s,  m,  h, wd, d,  m, y
        buf = bytearray([89, 69, 32, 3, 41, 5, 35])
        mock_readfrom_mem.return_value = buf

        self.assertEqual(ds1307.month, 5)

    @patch.object(I2C, 'readfrom_mem')
    def test_day(self, mock_readfrom_mem: MagicMock) -> None:
        """Test day property"""
        ds1307 = DS1307(i2c=self.i2c)

        # 2023-05-29T20:45:59
        #                0,  1,  2, 3,  4,  5, 6
        #                s,  m,  h, wd, d,  m, y
        buf = bytearray([89, 69, 32, 3, 41, 5, 35])
        mock_readfrom_mem.return_value = buf

        self.assertEqual(ds1307.day, 29)

    @patch.object(I2C, 'readfrom_mem')
    def test_hour(self, mock_readfrom_mem: MagicMock) -> None:
        """Test hour property"""
        ds1307 = DS1307(i2c=self.i2c)

        # 2023-05-29T20:45:59
        #                0,  1,  2, 3,  4,  5, 6
        #                s,  m,  h, wd, d,  m, y
        buf = bytearray([89, 69, 32, 3, 41, 5, 35])
        mock_readfrom_mem.return_value = buf

        self.assertEqual(ds1307.hour, 20)

    @patch.object(I2C, 'readfrom_mem')
    def test_minute(self, mock_readfrom_mem: MagicMock) -> None:
        """Test minute property"""
        ds1307 = DS1307(i2c=self.i2c)

        # 2023-05-29T20:45:59
        #                0,  1,  2, 3,  4,  5, 6
        #                s,  m,  h, wd, d,  m, y
        buf = bytearray([89, 69, 32, 3, 41, 5, 35])
        mock_readfrom_mem.return_value = buf

        self.assertEqual(ds1307.minute, 45)

    @patch.object(I2C, 'readfrom_mem')
    def test_second(self, mock_readfrom_mem: MagicMock) -> None:
        """Test second property"""
        ds1307 = DS1307(i2c=self.i2c)

        # 2023-05-29T20:45:59
        #                0,  1,  2, 3,  4,  5, 6
        #                s,  m,  h, wd, d,  m, y
        buf = bytearray([89, 69, 32, 3, 41, 5, 35])
        mock_readfrom_mem.return_value = buf

        self.assertEqual(ds1307.second, 59)

    @patch.object(I2C, 'readfrom_mem')
    def test_weekday(self, mock_readfrom_mem: MagicMock) -> None:
        """Test weekday property"""
        ds1307 = DS1307(i2c=self.i2c)

        # 2023-05-29T20:45:59
        #                0,  1,  2, 3,  4,  5, 6
        #                s,  m,  h, wd, d,  m, y
        buf = bytearray([89, 69, 32, 0, 41, 5, 35])
        mock_readfrom_mem.return_value = buf

        self.assertEqual(ds1307.weekday, 59)

    @patch.object(I2C, 'readfrom_mem')
    def test_yearday(self, mock_readfrom_mem: MagicMock) -> None:
        """Test yearday property"""
        ds1307 = DS1307(i2c=self.i2c)

        # 2023-05-29T20:45:59
        #                0,  1,  2, 3,  4,  5, 6
        #                s,  m,  h, wd, d,  m, y
        buf = bytearray([89, 69, 32, 0, 41, 5, 35])
        mock_readfrom_mem.return_value = buf

        self.assertEqual(ds1307.yearday, 149)

    @patch.object(I2C, 'readfrom_mem')
    def test_is_leap_year(self, mock_readfrom_mem: MagicMock) -> None:
        """Test leap year property"""
        ds1307 = DS1307(i2c=self.i2c)

        # 2023-05-29T20:45:59
        #                0,  1,  2, 3,  4,  5, 6
        #                s,  m,  h, wd, d,  m, y
        buf = bytearray([89, 69, 32, 0, 41, 5, 35])
        mock_readfrom_mem.return_value = buf

        self.assertEqual(ds1307.is_leap_year(2023), False)
        self.assertEqual(ds1307.is_leap_year(2024), True)

    def test_day_of_year(self) -> None:
        """Test day of year"""
        self.assertEqual(
            self.ds1307.day_of_year(year=2023, month=5, day=29),
            149
        )
        self.assertEqual(
            self.ds1307.day_of_year(year=2024, month=3, day=26),
            86
        )

    @patch.object(I2C, 'readfrom_mem')
    def test_halt(self, mock_readfrom_mem: MagicMock) -> None:
        """Test power up/down of RTC"""
        ds1307 = DS1307(i2c=self.i2c)

        self.assertFalse(ds1307.halt)

        for halt, buf, expectation in zip(
            [True, True, False, False],
            [
                # 2023-05-29T20:45:59
                [89, 69, 32, 0, 41, 5, 35],
                # 2023-05-29T20:46:00
                [0, 70, 32, 0, 41, 5, 35],
                # 2023-05-29T20:45:59
                [89, 69, 32, 0, 41, 5, 35],
                # 2023-05-29T20:46:00
                [0, 70, 32, 0, 41, 5, 35]
            ],
                [0xd9, 0x80, 0x59, 0x0]):
            with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
                mock_readfrom_mem.return_value = buf
                ds1307.halt = halt

                self.assertEqual(len(self._tracked_call_data), 1)
                self.assertEqual(
                    self._tracked_call_data[0]['args'],
                    (ds1307.addr, 0, bytearray([expectation]))
                )
                self._tracked_call_data = []

    def test_square_wave(self) -> None:
        """Test setting square wave output pin"""
        with self.assertRaises(ValueError) as context:
            self.ds1307.square_wave(sqw=2)

        self.assertEqual(
            str(context.exception),
            "Squarewave can be set to 0Hz, 1Hz, 4096Hz, 8192Hz or 32768Hz"
        )

        ds1307 = DS1307(i2c=self.i2c)

        for sqw, expectation in zip((0, 1, 4, 8, 32),
                                    (0x80, 0x90, 0x91, 0x92, 0x93)):
            with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
                ds1307.square_wave(sqw=sqw, out=1)

            self.assertEqual(len(self._tracked_call_data), 1)
            self.assertEqual(
                self._tracked_call_data[0]['args'],
                (ds1307.addr, 7, bytearray([expectation]))
            )
            self._tracked_call_data = []

    def test__dec_to_bcd(self) -> None:
        """Test converting decimal to BCD"""
        self.assertEqual(self.ds1307._dec_to_bcd(value=1), 1)
        self.assertEqual(self.ds1307._dec_to_bcd(value=5), 5)
        self.assertEqual(self.ds1307._dec_to_bcd(value=10), 16)
        self.assertEqual(self.ds1307._dec_to_bcd(value=11), 17)
        self.assertEqual(self.ds1307._dec_to_bcd(value=300), 480)

    def test__bcd_to_dec(self) -> None:
        """Test converting BCD to decimal"""
        self.assertEqual(self.ds1307._bcd_to_dec(value=1), 1)
        self.assertEqual(self.ds1307._bcd_to_dec(value=5), 5)
        self.assertEqual(self.ds1307._bcd_to_dec(value=26), 20)
        self.assertEqual(self.ds1307._bcd_to_dec(value=2023), 1267)

    def tearDown(self) -> None:
        """Run after every test method"""
        pass


if __name__ == '__main__':
    unittest.main()


ds1307/ds1307.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""
I2C RTC driver for DS1307

https://github.com/mcauser/micropython-tinyrtc-i2c

MIT License
Copyright (c) 2018 Mike Causer
Extended 2023 by brainelectronics

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

BCD Format:
https://en.wikipedia.org/wiki/Binary-coded_decimal
"""

# system packages
from machine import I2C
try:
    from micropython import const
except ImportError:
    def const(x):
        return x


class _Subscriptable():
    def __getitem__(self, item):
        return None


_subscriptable = _Subscriptable()

Optional = _subscriptable
Tuple = _subscriptable

DATETIME_REG = const(0)     # 0x00-0x06
CHIP_HALT = const(128)
CONTROL_REG = const(7)      # 0x07
RAM_REG = const(8)          # 0x08-0x3F


class DS1307(object):
    """Driver for the DS1307 RTC."""

    def __init__(self, addr=0x68, i2c: Optional[I2C] = None) -> None:
        """
        Constructs a new instance

        :param      addr:   The I2C bus address of the EEPROM
        :type       addr:   int
        :param      i2c:    I2C object
        :type       i2c:    I2C
        """
        self._addr = addr

        if i2c is None:
            # default assignment, check the docs
            self._i2c = I2C(0)
        else:
            self._i2c = i2c

        self._weekday_start = 0
        self._halt = False

    @property
    def addr(self) -> int:
        """
        Get the DS1307 I2C bus address

        :returns:   DS1307 I2C bus address
        :rtype:     int
        """
        return self._addr

    @property
    def weekday_start(self) -> int:
        """
        Get the start of the weekday

        :returns:   Weekday start
        :rtype:     int
        """
        return self._weekday_start

    @weekday_start.setter
    def weekday_start(self, value: int) -> None:
        """Set the start of the weekday"""
        if 0 <= value <= 6:
            self._weekday_start = value
        else:
            raise ValueError("Weekday can only be in range 0-6")

    @property
    def datetime(self) -> Tuple[int, int, int, int, int, int, int, int]:
        """
        Get the current datetime

        (2023, 4, 18, 0, 10, 34, 4, 108)
        y,     m,  d, h, m,  s, wd, yd

        :returns:   (year, month, day, hour, minute, second, weekday, yearday)
        :rtype:     Tuple[int, int, int, int, int, int, int, int]
        """
        buf = bytearray(7)

        buf = self._i2c.readfrom_mem(self._addr, DATETIME_REG, 7)

        year = self._bcd_to_dec(buf[6]) + 2000
        month = self._bcd_to_dec(buf[5])
        day = self._bcd_to_dec(buf[4])
        yearday = self.day_of_year(year=year, month=month, day=day)
        return (
            year,
            month,
            day,
            self._bcd_to_dec(buf[2]),           # hour
            self._bcd_to_dec(buf[1]),           # minute
            self._bcd_to_dec(buf[0] & 0x7F),    # second
            self._bcd_to_dec(buf[3] - self._weekday_start),     # weekday
            yearday,
        )

    @datetime.setter
    def datetime(self, datetime: Tuple[int, int, int, int, int, int, int, int]) -> None:    # noqa: E501
        """
        Set datetime

        (2023, 4, 18, 20, 23, 38, 1, 108) by time.gmtime(time.time())
        y,     m,  d,  h, min,sec,wday, yday
        0,     1,  2,  3,  4,   5,   6, 7

        :param      datetime:  The datetime
        :type       datetime:  Tuple[int, int, int, int, int, int, int, int]
        """
        buf = bytearray(7)

        # msb = CH, 1 = halt, 0 = go
        buf[0] = self._dec_to_bcd(datetime[5]) & 0x7F   # second
        buf[1] = self._dec_to_bcd(datetime[4])  # minute
        buf[2] = self._dec_to_bcd(datetime[3])  # hour
        buf[3] = self._dec_to_bcd(datetime[6] + self._weekday_start)
        buf[4] = self._dec_to_bcd(datetime[2])  # day
        buf[5] = self._dec_to_bcd(datetime[1])  # month
        buf[6] = self._dec_to_bcd(datetime[0] - 2000)   # year

        if (self._halt):
            buf[0] |= (1 << 7)

        self._i2c.writeto_mem(self._addr, DATETIME_REG, buf)

    @property
    def year(self) -> int:
        """
        Get the year from the RTC

        :returns:   Year of RTC
        :rtype:     int
        """
        return self.datetime[0]

    @property
    def month(self) -> int:
        """
        Get the month from the RTC

        :returns:   Month of RTC
        :rtype:     int
        """
        return self.datetime[1]

    @property
    def day(self) -> int:
        """
        Get the day from the RTC

        :returns:   Day of RTC
        :rtype:     int
        """
        return self.datetime[2]

    @property
    def hour(self) -> int:
        """
        Get the hour from the RTC

        :returns:   Hour of RTC
        :rtype:     int
        """
        return self.datetime[3]

    @property
    def minute(self) -> int:
        """
        Get the minute from the RTC

        :returns:   Minute of RTC
        :rtype:     int
        """
        return self.datetime[4]

    @property
    def second(self) -> int:
        """
        Get the second from the RTC

        :returns:   Second of RTC
        :rtype:     int
        """
        return self.datetime[5]

    @property
    def weekday(self) -> int:
        """
        Get the weekday from the RTC

        :returns:   Weekday of RTC
        :rtype:     int
        """
        return self.datetime[5]

    @property
    def yearday(self) -> int:
        """
        Get the yearday from the RTC

        :returns:   Yearday of RTC
        :rtype:     int
        """
        return self.datetime[7]

    def is_leap_year(self, year: int) -> bool:
        """
        Determines whether the specified year is a leap year.

        :param      year:  The year
        :type       year:  int

        :returns:   True if the specified year is leap year, False otherwise.
        :rtype:     bool
        """
        return (year % 4 == 0)

    def day_of_year(self, year: int, month: int, day: int) -> int:
        """
        Get the day of the year

        :param      year:   The year
        :type       year:   int
        :param      month:  The month
        :type       month:  int
        :param      day:    The day
        :type       day:    int

        :returns:   Day of the year, January 1 is day 1
        :rtype:     int
        """
        month_days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
        year -= 2000
        days16 = day

        for x in range(1, month):
            days16 += month_days[x - 1]

        if month >= 2 and self.is_leap_year(year=year):
            days16 += 1

        return days16

    @property
    def halt(self) -> bool:
        """
        Get status of DS1307

        :returns:   Status of DS1307
        :rtype:     bool
        """
        return self._halt

    @halt.setter
    def halt(self, value: bool = False) -> None:
        """
        Power up or power down the RTC oscillator

        :param      value:  The value
        :type       value:  bool
        """
        reg = self._i2c.readfrom_mem(self._addr, DATETIME_REG, 1)[0]

        if value:
            reg |= CHIP_HALT
        else:
            reg &= ~CHIP_HALT

        self._halt = bool(value)
        self._i2c.writeto_mem(self._addr, DATETIME_REG, bytearray([reg]))

    def square_wave(self, sqw: int = 0, out: int = 0) -> None:
        """
        Output square wave on pin SQ

        at 1Hz, 4.096kHz, 8.192kHz or 32.768kHz, or disable the oscillator
        and output logic level high/low.
        """
        if sqw not in (0, 1, 4, 8, 32):
            raise ValueError(
                "Squarewave can be set to 0Hz, 1Hz, {}Hz, {}Hz or {}Hz".format(
                    2 ** 12, 2 ** 13, 2 ** 15)
            )

        rs0 = 1 if sqw == 4 or sqw == 32 else 0
        rs1 = 1 if sqw == 8 or sqw == 32 else 0
        out = 1 if out > 0 else 0
        sqw = 1 if sqw > 0 else 0

        reg = rs0 | rs1 << 1 | sqw << 4 | out << 7

        self._i2c.writeto_mem(self._addr, CONTROL_REG, bytearray([reg]))

    def _dec_to_bcd(self, value: int) -> int:
        """
        Convert decimal to binary coded decimal (BCD) format

        :param      value:  The value
        :type       value:  int

        :returns:   Binary coded decimal (BCD) format
        :rtype:     int
        """
        return (value // 10) << 4 | (value % 10)

    def _bcd_to_dec(self, value: int) -> int:
        """
        Convert binary coded decimal (BCD) format to decimal

        :param      value:  The value
        :type       value:  int

        :returns:   Decimal value
        :rtype:     int
        """
        return ((value >> 4) * 10) + (value & 0x0F)


examples/main.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""I2C DS1307 showcase"""

from ds1307 import DS1307
from machine import I2C, Pin
from time import gmtime, sleep, time

# DS1307 on 0x68
I2C_ADDR = 0x68     # DEC 104, HEX 0x68

# define custom I2C interface, default is 'I2C(0)'
# check the docs of your device for further details and pin infos
# this are the pins for the Raspberry Pi Pico adapter board
i2c = I2C(0, scl=Pin(13), sda=Pin(12), freq=800000)
ds1307 = DS1307(addr=I2C_ADDR, i2c=i2c)

# get LCD infos/properties
print("DS1307 is on I2C address 0x{0:02x}".format(ds1307.addr))
print("Weekday start is {}".format(ds1307.weekday_start))

# get the current RTC time
print("Current RTC time: {}".format(ds1307.datetime))

# set the RTC time to the current system time
now = gmtime(time())
ds1307.datetime = now

# Print the date and time in ISO8601 format: 2023-04-18T21:14:22
print("Today is {:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}".format(
    ds1307.year, ds1307.month, ds1307.day,
    ds1307.hour, ds1307.minute, ds1307.second))

# check whether this year is a leap year
print("Is this year a leap year? {}".format(ds1307.is_leap_year(ds1307.year)))

# get the day of the year
print("Today is day {} of {}".format(
    ds1307.day_of_year(year=ds1307.year, month=ds1307.month, day=ds1307.day),
    ds1307.year))

# halt the oscillator
print("The oscillator is currently active at {}? {}".format(
    ds1307.datetime, ds1307.halt))
print("Halt the oscillator and wait for 5 seconds ...")
ds1307.halt = True
sleep(5)

print("Current RTC time: {}".format(ds1307.datetime))

print("Enable the oscillator and wait for 5 seconds ...")
ds1307.halt = False
sleep(5)
print("Current RTC time: {}".format(ds1307.datetime))

# control the squarewave pin SQ
print("Set square wave output to 1Hz and wait for 5 seconds ...")
ds1307.square_wave(sqw=1)
sleep(5)

print("Set square wave output to 4.096kHz and wait for 5 seconds ...")
ds1307.square_wave(sqw=4)
sleep(5)

print("Set square wave output to 8.192kHz and wait for 5 seconds ...")
ds1307.square_wave(sqw=8)
sleep(5)

print("Set square wave output to HIGH and wait for 5 seconds ...")
ds1307.square_wave(out=1)
sleep(5)

print("Set square wave output to LOW and wait for 5 seconds ...")
ds1307.square_wave(sqw=0)
sleep(5)
